swap/src/alice/swap.rs

```diff
@@ -393,67 +393,72 @@ pub async fn run_until(
                 )
                 .await
             }
-            AliceState::XmrLocked { state3 } => {
+            AliceState::XmrLocked { ref state3 } => {
                 // todo: match statement and wait for t1 can probably be expressed more cleanly
-                let (state, event_loop_handle) =
-                    match state3.current_epoch(bitcoin_wallet.as_ref()).await? {
-                        Epoch::T0 => {
-                            let wait_for_enc_sig = wait_for_bitcoin_encrypted_signature(
-                                &mut event_loop_handle,
-                                config.monero_max_finality_time,
-                            );
-                            let t1_timeout = state3.wait_for_t1(bitcoin_wallet.as_ref());
-
-                            tokio::select! {
-                                _ = t1_timeout => {
-                                    run_until(
-                                        AliceState::Cancelling { state3 },
-                                        is_target_state,
-                                        event_loop_handle,
-                                        bitcoin_wallet,
-                                        monero_wallet,
-                                        config,
-                                    )
-                                    .await
-                                }
-                                enc_sig = wait_for_enc_sig => {
-                                      run_until(
-                                            AliceState::EncSignLearned {
-                                                state3,
-                                                encrypted_signature: enc_sig?,
-                                            },
-                                            is_target_state,
-                                            event_loop_handle,
-                                            bitcoin_wallet,
-                                            monero_wallet,
-                                            config,
-                                        )
-                                        .await
-                                }
+                match state3.current_epoch(bitcoin_wallet.as_ref()).await? {
+                    Epoch::T0 => {
+                        let wait_for_enc_sig = wait_for_bitcoin_encrypted_signature(
+                            &mut event_loop_handle,
+                            config.monero_max_finality_time,
+                        );
+                        let t1_timeout = state3.wait_for_t1(bitcoin_wallet.as_ref());
+
+                        tokio::select! {
```

Comment:
In my try I replaced this with `futures::future::select`. It allows to return a value, the state, removing the triplicate db save and `run_until` call.